File: web/src/lib/serverState.test.ts
import { describe, it, expect } from 'vitest'
import {
  mapServerState,
  containerRows,
  failedStacks,
  containerStateVariant,
  displayName,
  healthVariant,
  versionLabel,
} from './serverState'

const camelCase = {
  serverId: 'srv-1',
  reportedAt: 1700000000000,
  stacks: [
    {
      stackId: 'st-a',
      stackName: 'web-stack',
      runningVersionId: 'ver-a1',
      runningVersion: 3,
      containers: [
        {
          id: 'c1',
          name: 'st-a-web-1234abcd',
          image: 'nginx:1.27',
          serviceName: 'web',
          state: 'running',
          status: 'Up 3 hours',
          health: 'healthy',
          restartCount: 2,
          startedAt: 1699999999000,
        },
      ],
    },
  ],
}

const snakeCase = {
  server_id: 'srv-1',
  reported_at: 1700000000000,
  stacks: [
    {
      stack_id: 'st-a',
      stack_name: 'web-stack',
      running_version_id: 'ver-a1',
      running_version: 3,
      containers: [
        {
          id: 'c1',
          name: 'st-a-web-1234abcd',
          image: 'nginx:1.27',
          service_name: 'web',
          state: 'running',
          status: 'Up 3 hours',
          health: 'healthy',
          restart_count: 2,
          started_at: 1699999999000,
        },
      ],
    },
  ],
}

describe('mapServerState', () => {
  it('maps a camelCase response', () => {
    const state = mapServerState(camelCase)
    expect(state.serverId).toBe('srv-1')
    expect(state.reportedAt).toBe(1700000000000)
    expect(state.stacks[0].runningVersion).toBe(3)
    expect(state.stacks[0].containers[0].serviceName).toBe('web')
    expect(state.stacks[0].containers[0].restartCount).toBe(2)
  })

  it('maps a snake_case response identically', () => {
    expect(mapServerState(snakeCase)).toEqual(mapServerState(camelCase))
  })

  it('denormalises the stack onto each container row', () => {
    const c = mapServerState(camelCase).stacks[0].containers[0]
    expect(c.stackId).toBe('st-a')
    expect(c.stackName).toBe('web-stack')
  })

  it('maps per-service errors in both casings', () => {
    const a = mapServerState({ stacks: [{ serviceErrors: [{ serviceName: 'api', error: 'boom' }] }] })
    const b = mapServerState({ stacks: [{ service_errors: [{ service_name: 'api', error: 'boom' }] }] })
    expect(a.stacks[0].serviceErrors).toEqual([{ serviceName: 'api', error: 'boom' }])
    expect(b).toEqual(a)
  })

  it('survives an empty or missing payload', () => {
    expect(mapServerState({})).toEqual({ serverId: '', reportedAt: 0, stacks: [] })
    expect(mapServerState(undefined)).toEqual({ serverId: '', reportedAt: 0, stacks: [] })
    expect(mapServerState({ stacks: null })).toEqual({ serverId: '', reportedAt: 0, stacks: [] })
  })

  it('defaults missing container fields instead of rendering undefined', () => {
    const c = mapServerState({ stacks: [{ stack_id: 'st-a', containers: [{ id: 'c1' }] }] })
      .stacks[0].containers[0]
    expect(c).toMatchObject({ name: '', image: '', health: '', restartCount: 0, startedAt: 0 })
  })
})

describe('containerRows', () => {
  it('flattens stacks and sorts by stack then service', () => {
    const state = mapServerState({
      stacks: [
        { stack_id: 'st-b', stack_name: 'beta', containers: [{ id: 'c3', service_name: 'api' }] },
        {
          stack_id: 'st-a',
          stack_name: 'alpha',
          containers: [
            { id: 'c2', service_name: 'web' },
            { id: 'c1', service_name: 'db' },
          ],
        },
      ],
    })
    expect(containerRows(state).map(c => c.id)).toEqual(['c1', 'c2', 'c3'])
  })

  it('returns nothing for an undefined state', () => {
    expect(containerRows(undefined)).toEqual([])
  })
})

describe('failedStacks', () => {
  it('returns only stacks with a reconcile error', () => {
    const state = mapServerState({
      stacks: [{ stack_id: 'st-a' }, { stack_id: 'st-b', error: 'converge failed' }],
    })
    expect(failedStacks(state).map(s => s.stackId)).toEqual(['st-b'])
  })
})

describe('displayName', () => {
  const row = (name: string, stackId = 'st-a', id = 'abcdef1234567890') =>
    mapServerState({ stacks: [{ stack_id: stackId, containers: [{ id, name }] }] })
      .stacks[0].containers[0]

  it('drops the repeated stack-id prefix', () => {
    expect(displayName(row('st-a-web-1234abcd'))).toBe('web-1234abcd')
  })

  it('leaves a name that does not carry the prefix untouched', () => {
    expect(displayName(row('my-container'))).toBe('my-container')
  })

  it('falls back to a short container id when there is no name', () => {
    expect(displayName(row(''))).toBe('abcdef123456')
  })
})

describe('containerStateVariant', () => {
  it('maps docker states to badge variants', () => {
    expect(containerStateVariant('running')).toBe('online')
    expect(containerStateVariant('exited')).toBe('offline')
    expect(containerStateVariant('restarting')).toBe('warn')
    expect(containerStateVariant('')).toBe('warn')
  })
})

describe('healthVariant', () => {
  it('maps health to badge variants, with no healthcheck as default', () => {
    expect(healthVariant('healthy')).toBe('online')
    expect(healthVariant('unhealthy')).toBe('error')
    expect(healthVariant('starting')).toBe('warn')
    expect(healthVariant('')).toBe('default')
  })
})

describe('versionLabel', () => {
  it('prefers the resolved numeric version', () => {
    const [stack] = mapServerState(camelCase).stacks
    expect(versionLabel(stack)).toBe('v3')
  })

  it('falls back to a short version id when the Master could not resolve it', () => {
    const [stack] = mapServerState({ stacks: [{ running_version_id: 'abcdef1234567890' }] }).stacks
    expect(versionLabel(stack)).toBe('abcdef12')
  })

  it('shows a dash for a stack with no unanimous version', () => {
    const [stack] = mapServerState({ stacks: [{ stack_id: 'st-a' }] }).stacks
    expect(versionLabel(stack)).toBe('—')
  })
})

File: internal/master/api/serverstate_test.go
package api

import (
	"testing"

	orkestraV1 "github.com/heckertobias/orkestra/internal/shared/gen/orkestra/v1"
)

func testReport() *orkestraV1.StatusReport {
	return &orkestraV1.StatusReport{
		ReportedAtMs: 1700000000000,
		Stacks: []*orkestraV1.StackStatus{
			{
				StackId:        "stack-a",
				RunningVersion: "ver-a1",
				Containers: []*orkestraV1.ContainerStatus{
					{
						ContainerId:  "c1",
						Name:         "stack-a-web-1234abcd",
						Image:        "nginx:1.27",
						ServiceName:  "web",
						State:        "running",
						Status:       "Up 3 hours",
						Health:       "healthy",
						RestartCount: 2,
						StartedAtMs:  1699999999000,
					},
				},
			},
			{
				StackId: "stack-b",
				Error:   "converge stack-b: 1 service(s) failed",
				ServiceErrors: []*orkestraV1.ServiceError{
					{ServiceName: "api", Error: "pull access denied"},
				},
			},
		},
	}
}

func TestServerStateFromReportConvertsContainers(t *testing.T) {
	state := serverStateFromReport("srv-1", testReport(), nil)

	if state.ServerId != "srv-1" || state.ReportedAt != 1700000000000 {
		t.Fatalf("server id/timestamp not carried through: %+v", state)
	}
	if len(state.Stacks) != 2 {
		t.Fatalf("expected 2 stacks, got %d", len(state.Stacks))
	}

	a := state.Stacks[0]
	if a.RunningVersionId != "ver-a1" {
		t.Fatalf("running version id = %q, want ver-a1", a.RunningVersionId)
	}
	// The numeric version is resolved from the database, not from the report.
	if a.RunningVersion != 0 {
		t.Fatalf("numeric version must stay unresolved here, got %d", a.RunningVersion)
	}
	if len(a.Containers) != 1 {
		t.Fatalf("expected 1 container, got %d", len(a.Containers))
	}
	c := a.Containers[0]
	if c.Id != "c1" || c.Name != "stack-a-web-1234abcd" || c.Image != "nginx:1.27" {
		t.Fatalf("container identity not carried through: %+v", c)
	}
	if c.ServiceName != "web" || c.State != "running" || c.Status != "Up 3 hours" {
		t.Fatalf("container state not carried through: %+v", c)
	}
	if c.Health != "healthy" || c.RestartCount != 2 || c.StartedAt != 1699999999000 {
		t.Fatalf("health/restarts/start time not carried through: %+v", c)
	}
}

func TestServerStateFromReportCarriesServiceErrors(t *testing.T) {
	state := serverStateFromReport("srv-1", testReport(), nil)

	b := state.Stacks[1]
	if b.Error == "" {
		t.Fatal("stack-level error must be carried through")
	}
	if len(b.ServiceErrors) != 1 || b.ServiceErrors[0].ServiceName != "api" {
		t.Fatalf("per-service errors must be carried through, got %v", b.ServiceErrors)
	}
	if len(b.Containers) != 0 {
		t.Fatalf("a stack with no containers must report none, got %d", len(b.Containers))
	}
}

func TestServerStateFromReportFiltersStacksTheCallerCannotView(t *testing.T) {
	// A stack-scoped binding must not leak the inventory of the other stacks on that server.
	state := serverStateFromReport("srv-1", testReport(), func(stackID string) bool {
		return stackID == "stack-a"
	})

	if len(state.Stacks) != 1 || state.Stacks[0].StackId != "stack-a" {
		t.Fatalf("expected only stack-a, got %v", state.Stacks)
	}
}

func TestServerStateFromReportHandlesEmptyReport(t *testing.T) {
	state := serverStateFromReport("srv-1", &orkestraV1.StatusReport{}, nil)

	if state.ServerId != "srv-1" || len(state.Stacks) != 0 || state.ReportedAt != 0 {
		t.Fatalf("empty report must convert to an empty state, got %+v", state)
	}
}
